### ISSUE TITLE
URLS from Holdingsand bib records

### ISSUE THREAD

**Issue by patrickzurek** [eXtensibleCatalog/xcmetadataservicestoolkit#636]
JIRA issue created by: rcook
Originally opened: 2011-06-21 04:01 PM

Issue body:
 (nt)

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2011-06-21 04:03 PM

Comment body:

Currently Drupal Toolkit is using urls that originate in the bib records and it should use the urls from the holdings.  I believe that a change to a service is required before this can happend.

Once spec'd out, please point to the requriements and then can reassign.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2011-09-23 08:40 PM

Comment body:

This requires both new schema elements and changes to Transformation.  The new schema elements are now posted to Docushare (document probably needs more work and discussion before posting publicly)
http://docushare.lib.rochester.edu/docushare/dsweb/View/Collection-5823

and the Transformation Service mapping changes are in Google Docs and linked from the wiki:

https://docs.google.com/spreadsheet/ccc?key=0ArAewjrTTKSUdFVwODhobWl5Q3RCdWl0Z1RkdFZLUXc&hl=en_US

This should be all ready to go

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-13 10:01 PM

Comment body:

Chris....can you some thought to this?  This one turns out to be pretty important at UR and am wondering if your folks also have this need (using the urls in holdings records as first choice over using the bibs.  This is all about the 856 in the records.

Peter...when this is done, you will end up with potentially 2 different data elements, same name, but in the holding and the manifestation.

If CARLI has this need I would like to move this one into a higher priority.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-19 05:02 PM

Comment body:

Chris...thoughts on this.  Trying to figure out your next work, but wanted you input on this.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-19 05:04 PM

Comment body:

A little vague for me.  I'm not sure what's needed.  Can someone give examples?  Something concrete (not a link to a general google doc)?  Thanks, Chris

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-19 05:10 PM

Comment body:

This record,
http://128.174.70.254/dtmilestone3/node/1822040?hit=1&caller=xc-search
has a url in the 852 of the holdings.  That links it to http://www.lib.rochester.edu/index.cfm?page=1127 (you probably can't see this being off campus)

The url is in the holdings record, not the bib.  So in Drupal, we have no link to display to the user.

I think the changes for Norm and the changes needed to support the extraction of holdings from bib records is next up, but wanted to check on this one, as it seems it could be pressing as well.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-19 05:15 PM

Comment body:

I'll study the issue and get back to you...
However, before doing so, I have a question (I'm not too familiar with the XC Drupal interface backend yet, so please bear with me):

Isn't Drupal harvesting XC schema (FRBR)?  And aren't all the holdings available to it already?

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-19 05:21 PM

Comment body:

Yes, however Drupal has the FRBR holdings records, and they only contain the data elements that the XC Transformation service puts into it.  So we need to add new data elements to the Holdings schema to hold the urls and the service needs to populate it.

Then later, Drupal needs some changes to make use of the newly available data.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-19 05:24 PM

Comment body:

I get it now.

Thanks for being patient.  I'm not well-versed on the frbr elements yet.

I think I have enough to go on now...

I'll ask questions if I have any (more) once I've had time to look things over.

Chris

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-19 05:28 PM

Comment body:

Chris, this is not officially your next work item yet, it is not yet in your working milestone 1.3.   I am trying to figure out if it is a concern for CARLI or not.  The work in the current 1.3 milestone is next up, and I am trying to figure out if this goes into the 1.3 list or not and how to prioritize it.

Jennifer, any input?

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2012-01-19 05:31 PM

Comment body:

The URL in the holdings record was a UR shortcut (that I much regret) and which may not be an issue for CARLI.  This is a legal practice in MARC, but not one that a lot of libraries used.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-19 05:33 PM

Comment body:

I believe some of our schools do embed links in their holdings (either in addition to, or instead of, I'm not sure).  My guess is that it does affect CARLI and is probably something we'd want.  I'll forward the question along to Gordon Fellows.  He'd have a better grasp of how frequent this scenario is and what sense of priority we might want to give it.  Chris

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-19 07:03 PM

Comment body:

Yes, it turns out this occurs often in CARLI's data.  I believe we'd also put a high priority on it.  I could begin work on this now if you think that's a good plan.  Chris

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-19 07:17 PM

Comment body:

I think that is fine.  It is, I think, a fairly straight forward fix, allowing you more exposure to changing the services before digging into the "harder" issues in the rest of 1.3 Milestone.   So I am OK with this.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-19 07:24 PM

Comment body:

Doing a notify to Peter, Dave and Jennifer.

I don't think we need anything else in the MST side (e.g. for config options).  The Transform will create the data elements if the tags are there (or do we need an on/off option?).  Then Drupal will later need to handle the new data elements (e.g. that Drupal needs to have config options about which urls to give priority to, those from holdings or from other FRBR levels).  If they just appear, it should not break anything, right Peter?  Seems we talked about this a long time ago but wanted to check.

Peter...I think you can do this easily, as you already handle data elements that can exist in both manifestion and holdings, like the XCrecordid.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-20 07:46 PM

Comment body:

In trunk, rev. 2912

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-20 07:56 PM

Comment body:

Chris, awesome.  

John, how do I test, can you set something up, maybe with the example I gave of Susan B Anthony papers, then I think Jennifer to approve the output?  And John, you can then add a test case for this to the release test set.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-20 08:12 PM

Comment body:

Attached are my input and output MockHarvest I used to test this.  It includes two bibs and a holding record.  One bib has an 856 link, and one holding record has an 856 link also.  The other bib is the "parent" record of the holding record.

Here is the relevant XC record output:

    <record>
      <header>
        <identifier>oai:mst.rochester.edu:MetadataServicesToolkit/marctoxctransformation/10</identifier>
        <datestamp>2012-01-20T20:07:20Z</datestamp>
        <setSpec>marctoxctransformation</setSpec>
      </header>
      <metadata>
        <xc:frbr xmlns:xc="http://www.extensiblecatalog.info/Elements" xmlns:dcterms="http://purl.org/dc/terms/" xmlns:rdarole="http://rdvocab.info/roles" xmlns:rdvocab="http://rdvocab.info/Elements">
          <xc:entity id="oai:mst.rochester.edu:MetadataServicesToolkit/marctoxctransformation/10" type="holdings">
            <xc:location>RBManuscrp</xc:location>
            <xc:callNumber>Manuscript Collection A.A62</xc:callNumber>
            <dcterms:relation>http://www.library.rochester.edu/index.cfm?page=1127 Register (List of contents)</dcterms:relation>
            <xc:recordID type="NRU">2185970</xc:recordID>
            <xc:manifestationHeld>oai:mst.rochester.edu:MetadataServicesToolkit/marctoxctransformation/7</xc:manifestationHeld>
          </xc:entity>
        </xc:frbr>
      </metadata>
    </record>


It's not an exhaustive test, but then again the code I added is identical to the bib processing of 856 field... pretty simple.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2012-01-20 08:55 PM

Comment body:

Chris, can you check in your mock harvest files?  Add in a new directory, call it whatever you want, thanks.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-20 09:00 PM

Comment body:

I added "fb699" directory.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-25 06:56 PM

Comment body:

I don't think this is correct.  Jennifer, please chime in here.

Jennifer, you can look at data here http://xc.ilcso.uiuc.edu:8080/MetadataServicesToolkit/st/allRepository.action

Below is from the MARC record.

â€¡u http://www.library.rochester.edu/index.cfm?page=1127 
â€¡3 Register 
â€¡z (List of contents) 

Below is the FRBR holdings record xml.  First, is the dcterm correct?  Chris is using <dcterms:relation> but should it be <dcterms:identifer>? 

Second, Jennifer, should this contain all of the 852 parts, or just the 
â€¡u http://www.library.rochester.edu/index.cfm?page=1127 part?


<xc:frbr xmlns:xc="http://www.extensiblecatalog.info/Elements" xmlns:dcterms="http://purl.org/dc/terms/" xmlns:rdarole="http://rdvocab.info/roles" xmlns:rdvocab="http://rdvocab.info/Elements" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <xc:entity id="oai:mst.rochester.edu:MetadataServicesToolkit/marctoxctransformation/19580895" type="holdings">
    <xc:location>RBManuscrp</xc:location>
    <xc:callNumber>Manuscript Collection A.A62</xc:callNumber>
    <dcterms:relation>http://www.library.rochester.edu/index.cfm?page=1127 Register (List of contents)</dcterms:relation>
    <xc:recordID type="NRU">2185970</xc:recordID>
    <xc:manifestationHeld>oai:mst.rochester.edu:MetadataServicesToolkit/marctoxctransformation/19580892</xc:manifestationHeld>
  </xc:entity>
</xc:frbr>

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-25 06:58 PM

Comment body:

Jennifer, Please look at several example of holdings records so we can identify all issues at one time so Chris can fix them all.  I would like to not piecemeal fix this too many times.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-25 07:02 PM

Comment body:

I meant 856 below, sorry.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-25 07:16 PM

Comment body:

Here's the 856 for the holding record ( http://xc.ilcso.uiuc.edu:8080/MetadataServicesToolkit/pub/marcnormalization/oaiRepository?verb=GetRecord&metadataPrefix=marc21&identifier=oai:mst.rochester.edu:MetadataServicesToolkit/marcnormalization/9737674 ):


<marc:datafield ind1="4" ind2="2" tag="856">
    <marc:subfield code="u">http://www.library.rochester.edu/index.cfm?page=1127</marc:subfield>
    <marc:subfield code="3">Register</marc:subfield>
    <marc:subfield code="z">(List of contents)</marc:subfield>
  </marc:datafield>


Here is the logic I am using on these 856s ( http://code.google.com/p/xcmetadataservicestoolkit/source/browse/trunk/mst-service/impl/src/java/xc/mst/services/impl/service/SolrTransformationService.java?spec=svn2912&r=2912#5990 ):

Note: I removed some logging code:

                // If 2nd indicator is 0, blank, or 8, map to manifestation
                // level dcterms:identifier
                if (ind2 == ' ' || ind2 == '0' || ind2 == '8') {

                    // Create an dcterms:identifier based on the 856
                    // abcdfhijklmnopqrstuvwxyz23 values
                    getXCRecordService().addElement(transformInto,
                            "identifier", value.trim(),
                            AggregateXCRecord.DCTERMS_NAMESPACE, attributes,
                            FrbrLevel.HOLDINGS);


	            } else if (ind2 == '1') {
	                // Create a dcterms:hasVersion based on the 856
	                // abcdfhijklmnopqrstuvwxyz23 values
	                getXCRecordService().addElement(transformInto,
	                        "hasVersion", value.trim(),
	                        AggregateXCRecord.DCTERMS_NAMESPACE, attributes,
	                        FrbrLevel.HOLDINGS);


	            } else if (ind2 == '2') {
		                // Create a dc:relation based on the 856
	                // abcdfhijklmnopqrstuvwxyz23 values
	                getXCRecordService().addElement(transformInto, "relation",
	                        value.trim(), AggregateXCRecord.DCTERMS_NAMESPACE,
	                        attributes, FrbrLevel.HOLDINGS);
	            }
            }


This is the same logic used for bibs too:

http://code.google.com/p/xcmetadataservicestoolkit/source/browse/trunk/mst-service/impl/src/java/xc/mst/services/impl/service/SolrTransformationService.java?spec=svn2912&r=2912#4949

Since this particular 856 does not have ind2 set to blank, 0, or 8, it is not a dcterms:identifier.  And since ind2 is a 2, it is therefore a dcterms:relation.

Unless, of course, I misread the spec...

Chris

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2012-01-27 04:23 PM

Comment body:

The URL in this example is NOT for a version of the resource (in which case it would get mapped to dcterms:identifier) but to a related resource, which is the register of what is in the collection of papers.  So Chris has implemented the spec correctly in terms of which element to map things to.

The problem with 856 fields like this is that they contain other data in addition to the URL, and figuring out how to display that in Drupal is tricky. Sometimes the text to display can be in a $3, like in this record.  I had thought that I had written the spec to include mapping that, but the spec actually leaves that out for both bibs and holdings (not sure why I did that, so, Chris, don't fix the code until we work through this more thoroughly!) In most cases a library will want to display $3 since it will actually tell users what the related resource is that the link represents.  In other cases, libraries have used the $z to determine what is displayed, instead of displaying the URL text.  That occurs in this example too.  But displaying info. from the $z (List of contents) may just look weird in Drupal unless Peter implements something special to control what displays - not sure if he has done that for bibs or not.  And I believe the text will just display in the order that the subfields were in the original MARC record, so it will be random, i.e. any one of these is possible unless Drupal does something special with it: 

www.XXXX.yyy Finding Aid
or
Finding Aid www.XXX.yyy
or 
www.XXX.yyy Click here for resource
or 
UR Access Only www.XXX.yyy

One possibility is to NOT map the $z so we won't have the messy display, although libraries would then lose the display text for the URL. Perhaps CARLI librarians could weigh in on how they feel about that.  For the UR, sometimes that text is "Access Limited to the University of Rochester" so I think we'd want to keep it.

We should probably also alter the spec to INCLUDE the $3, which is the way that Chris implemented it for holdings - we should make sure bibs work the same way, as I think that was a mistake on my part to leave it out.  That would be the easiest solution.  It's a bit messy, but at least no data is lost.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-02-08 07:55 PM

Comment body:

Jennifer, Is this done as is and able to go in next release?

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-02-20 05:34 PM

Comment body:

Not sure what is current status.  Chris, this is coded correct?   And will be in the next release?

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-02-20 07:27 PM

Comment body:

Randy,

I asked colleagues of mine for discussion regarding this matter.  I'll share this info once it becomes available.  

I have submitted this code into subversion trunk on Jan. 20, 2012 (rev. 2911).  Yes, it will go into the next release as per the current spec is defined.  If we wish to add to the spec, then I can make appropriate code changes then.  I'll leave this status to 'Active' until we (both CARLI and Jennifer) decide what, if anything, needs to be added or changed.

Chris

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-02-21 07:31 PM

Comment body:

My colleague Jessica pointed out that we do wish to include the $3.  And, btw, it is already being included today.  And, yes, we'd also like to keep the $z.

Here is more information from Jessica regarding the display:

"For the record, we spent a lot of time with VuFind defining how we wanted the 856 $3, $u, $y, and $z to display.  (Voyager does such a bad job that we were thrilled to finally have some control over it!)  We implemented the following in an attempt to follow the MARC standards for how the subfields are to be used balanced with the data that are frequently found in them. Here the â€œhyperlinkâ€� is always the URL data from the $u, and whatever is hyperlinked always displays in the OPAC:
If the 856 contains $u only: the $u will be hyperlinked.
If the 856 contains $u and $y: the $y will be hyperlinked.
If the 856 contains $u and $z: the $u will be hyperlinked and the $z will display as text.
If the 856 contains $u and $3: the $3 will be hyperlinked.
If the 856 contains $u, $3 and $y: the $3 and the $y will be hyperlinked, in that order, separated by a space.
If the 856 contains $u, $3 and $z: the $3 will be hyperlinked and the $z will display as text.
If the 856 contains $u, $y and $z: the $y will be hyperlinked and the $z will display as text.
If the 856 contains $u, $3, $y and $z: the $3 and $y will be hyperlinked, in that order, with a space between them, and the $z will display as text.
Any 856 $zâ€™s display as plain text after the hyperlink.
856 $z is repeatable, so if there is more than one, insert a space between each.  (Looking back, we probably should have put some kind of punctuation between repeated $zâ€™s as well.)
856 $y is technically repeatable, but it isnâ€™t clear why more than one would be needed, so currently we only display the first $y found in the field.
856 $u is technically repeatable, but our consortial cataloging guidelines allow only one URL ($u) per 856 field, so weâ€™ve never worried about this.
In short, both the 856 $3 and the 856 $z have important data that we would want to display in the XC OPAC. 
"

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-02-21 07:32 PM

Comment body:

Seems the MST part of this is complete.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2012-02-29 02:39 PM

Comment body:

regression test
ant -Dtest=MockHarvest -Dtest.folder=orig_186 test;
fails because needs to be updated to account for this change

dcterms:hasVersion

added to some of the expected files.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-03-08 04:05 PM

Comment body:

in release 1.3

**Comment by patrickzurek** [comment]
Issue resolved: 2012-03-08 04:05 PM
